quad/src/qag.rs

```diff
@@ -133,15 +133,51 @@ impl Qag {
     {
         if b == f64::INFINITY && a.is_finite() {
             let f2 = |x: f64| semi_infinite_function(&f, x, a, b);
-            return self.qintegrate(&f2, 0.0, 1.0, epsabs, epsrel);
+            let mut points = self.points.clone();
+            points.sort_by(|a, b| a.partial_cmp(b).unwrap());
+            let mut points_transformed = vec![0.0; 0];
+            for point in points {
+                points_transformed.push(1.0 / (point - a + 1.0));
+            }
+            let qag = Qag {
+                key: self.key,
+                limit: self.limit,
+                points: points_transformed,
+                more_info: self.more_info,
+            };
+            return qag.qintegrate(&f2, 0.0, 1.0, epsabs, epsrel);
         }
         if a == f64::NEG_INFINITY && b.is_finite() {
             let f2 = |x: f64| semi_infinite_function(&f, x, b, a);
-            return self.qintegrate(&f2, 0.0, 1.0, epsabs, epsrel);
+            let mut points = self.points.clone();
+            points.sort_by(|a, b| a.partial_cmp(b).unwrap());
+            let mut points_transformed = vec![0.0; 0];
+            for point in points {
+                points_transformed.push(1.0 / (b - point + 1.0));
+            }
+            let qag = Qag {
+                key: self.key,
+                limit: self.limit,
+                points: points_transformed,
+                more_info: self.more_info,
+            };
+            return qag.qintegrate(&f2, 0.0, 1.0, epsabs, epsrel);
         }
         if a == f64::NEG_INFINITY && b == f64::INFINITY {
             let f2 = |x: f64| double_infinite_function(&f, x);
-            return self.qintegrate(&f2, -1.0, 1.0, epsabs, epsrel);
+            let mut points = self.points.clone();
+            points.sort_by(|a, b| a.partial_cmp(b).unwrap());
+            let mut points_transformed = vec![0.0; 0];
+            for point in points {
+                points_transformed.push(point.signum() / (point.abs() + 1.0));
+            }
+            let qag = Qag {
+                key: self.key,
+                limit: self.limit,
+                points: points_transformed,
+                more_info: self.more_info,
+            };
+            return qag.qintegrate(&f2, -1.0, 1.0, epsabs, epsrel);
```

Comment:
This piece of code is repeated three times: just split into its own function, and pass the different element, i.e. the closure:
```rs
let qag = qag_from_points(|point, _a, _b| point.signum() / (point.abs() + 1.0))
```

Reply:
I tried to reduce the code repetition, do you think that the way I did it is enough?

Reply:
Yes, that's fine.